modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -1724,20 +1635,17 @@ private CompletableFuture<Object> applyFinishCommand(
      * @param request Transaction cleanup request.
      * @return CompletableFuture of void.
      */
-    // TODO: need to properly handle primary replica changes https://issues.apache.org/jira/browse/IGNITE-17615
     private CompletableFuture<Void> processWriteIntentSwitchAction(WriteIntentSwitchReplicaRequest request) {
-        try {
-            closeAllTransactionCursors(request.txId());
-        } catch (Exception e) {
-            return failedFuture(e);
-        }
-
-        TxState txState = request.commit() ? COMMITTED : ABORTED;
-
-        markFinished(request.txId(), txState, request.commitTimestamp());
+        markFinished(request.txId(), request.commit() ? COMMITTED : ABORTED, request.commitTimestamp());
 
         return awaitCleanupReadyFutures(request.txId(), request.commit())
                 .thenCompose(res -> {
+                    try {
+                        closeAllTransactionCursors(request.txId());
+                    } catch (Exception e) {
+                        return failedFuture(e);
+                    }
```

Comment:
I am not sure that we should cancel the write intent switch operation if we failed to close some cursor